AK/IpAddressCidr.h

```diff
@@ -0,0 +1,175 @@
+    Result<void, IPAddressCidrError> set_length(u32 length)
+    {
+        if (length > AddressFamily::MAX_LENGTH)
+            return IPAddressCidrError::CidrTooLong;
+
+        m_length = length;
+        return Result<void, IPAddressCidrError>();
+    }
+
+    static Result<AddressFamily, IPAddressCidrError> from_string(StringView string)
+    {
+        Vector<StringView> const parts = string.split_view('/');
+
+        if (parts.size() != 2)
+            return IPAddressCidrError::StringParsingFailed;
+
+        Optional ip_address = AddressTraits<AddressFamily>::IpAddress::from_string(parts[0]);
+        if (!ip_address.has_value())
+            return IPAddressCidrError::StringParsingFailed;
+
+        Optional<u8> length = parts[1].to_number<u8>();
+        if (!length.has_value())
+            return IPAddressCidrError::StringParsingFailed;
+
+        return IPAddressCidr::Create(ip_address.value(), length.value());
+    }
+
+#ifdef KERNEL
+    ErrorOr<NonnullOwnPtr<Kernel::KString>> to_string() const
+#else
+    ErrorOr<String> to_string() const
+#endif
+    {
+        StringBuilder builder;
+
+        TRY(builder.try_append(TRY(m_address.to_string())));
+        TRY(builder.try_append('/'));
+        TRY(builder.try_appendff("{}", m_length));
+
+#ifdef KERNEL
+        return Kernel::KString::try_create(builder.string_view());
+#else
+        return builder.to_string();
+#endif
+    }
+
+    constexpr bool operator==(IPAddressCidr const& other) const = default;
+    constexpr bool operator!=(IPAddressCidr const& other) const = default;
+
+protected:
+    constexpr IPAddressCidr(AddressTraits<AddressFamily>::IpAddress address, u8 length)
+    {
+        m_address = address;
+        m_length = length;
+    }
+
+private:
+    AddressTraits<AddressFamily>::IpAddress m_address;
+    u8 m_length;
+};
+
+template<>
+class AddressTraits<IPv4AddressCidr> {
+public:
+    using IpAddress = IPv4Address;
+};
+
+class [[gnu::packed]] IPv4AddressCidr : public IPAddressCidr<IPv4AddressCidr> {
+public:
+    IPv4AddressCidr(IPv4Address address, u8 length)
+        : IPAddressCidr(address, length)
+    {
+    }
+
+    IPv4Address netmask() const
+    {
+        u8 free_bits = MAX_LENGTH - length();
+        u32 mask = -1;
+        return IPv4Address(mask >> free_bits);
+    }
+
+    IPv4Address first_address_of_subnet() const
+    {
+        u32 mask = netmask().to_u32();
+        return IPv4Address(ip_address().to_u32() & mask);
+    }
+
+    IPv4Address last_address_of_subnet() const
+    {
+        u32 mask = netmask().to_u32();
+        u32 first = ip_address().to_u32() & mask;
+        return IPv4Address(first | ~mask);
+    }
+
+    bool contains(IPv4Address other) const
+    {
+        IPv4AddressCidr other_cidr = IPv4AddressCidr::Create(other, length()).value();
+        return first_address_of_subnet() == other_cidr.first_address_of_subnet();
+    }
+
+    static u8 const MAX_LENGTH = 32;
```

Comment:
this is `constexpr` (which will allow you to make a lot of other stuff constexpr)